app/src/main/java/tech/jpco/qen/view/QenPage.kt

```diff
@@ -0,0 +1,189 @@
+    private lateinit var bufferBitmap: Bitmap
+    private lateinit var bufferCanvas: Canvas
+    private val paint = Paint()
+    private val hypoteneuse
+        get() = sqrt((height * height + width * width).toDouble())
+            .also { if (it == 0.0) iLogger("hypoteneuse was zero!") }
+    private var minTouchDistance: Float by Delegates.notNull()
+
+    override fun onDraw(canvas: Canvas?) {
+        super.onDraw(canvas)
+        canvas?.drawBitmap(bufferBitmap, 0f, 0f, paint)
+    }
+
+    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
+        Log.d(TAG, "onSizeChanged called")
+        ar = w.toFloat() / h
+        paint.strokeWidth = (hypoteneuse / 300).toFloat()
+        minTouchDistance = (15 /*px*/ / hypoteneuse).toFloat()
+        //TODO make this resilient/persistent
+        bufferBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
+        bufferCanvas = Canvas(bufferBitmap)
+        super.onSizeChanged(w, h, oldw, oldh)
+    }
+
+
+    val touchStream: Observable<DrawPoint> = touches()
+        .filterForActionAndTime(33)
+        .map {
+            val (normX, normY) = normalize(it.x, it.y)
+            val type = when (it.action) {
+                MotionEvent.ACTION_DOWN -> TouchEventType.TouchDown
+                MotionEvent.ACTION_UP -> TouchEventType.TouchUp
+                MotionEvent.ACTION_MOVE -> TouchEventType.TouchMove
+                else -> throw IllegalArgumentException(
+                    "Tried to turn a MotionEvent other than Down, Up, or " +
+                            "Move into a DrawPoint!"
+                )
+            }
+            DrawPoint(normX, normY, type)
+        }
+        .filterForMinDistance { x, y -> abs(x) + abs(y) < minTouchDistance }
+
+    val arStream: Observable<Float> = layoutChanges().map { ar }
+
+    fun observeTouchStreamList(touchStreams: List<Observable<DrawPoint>>): List<Disposable> {
+        var counter = 0
+        return touchStreams.map { userStream ->
+            val localPaint = newPaint(counter++)
+            userStream.scan(
+                Pair<DrawPoint?, DrawPoint>(null, DrawPoint(Float.NaN, Float.NaN))
+            ) { pair, new ->
+                if (pair.second.x.isNaN() || new.type == TouchEventType.TouchDown)
+                    return@scan null to new
+                pair.second to new
+            }
+                .log("paired touchstream", this)
+                .filter { it.first != null }
+                .map { it as Pair<DrawPoint, DrawPoint> }
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe {
+                    iLogger("drawing ${it.second}")
+                    it.drawLine(localPaint)
+                    invalidate()
+                }
+        }
+    }
+
+    private val goldenConjugate = (2 / (sqrt(5f) + 1))
+    private val paintSequenceList = mutableListOf<Paint>()
+    private val newPaint: (Int) -> Paint = { counter ->
```

Comment:
This looks like it's being called on a code hotpath (inside `touchStreams.map { ... }`) so it might be better to promote this to being a full function.